corehq/motech/value_source.py

```diff
@@ -253,160 +246,186 @@ class ConstantValue(ConstantString):
     A reference to a location metadata value. The location is the form
     user's location, or the first ancestor location of the form user
     where the metadata value is set.
     """
-    location_field = StringProperty()
+    form_user_ancestor_location_field: str
 
-    def _get_commcare_value(self, case_trigger_info):
+    @classmethod
+    def wrap(cls, data):
+        if "location_field" in data:
+            data["form_user_ancestor_location_field"] = data.pop("location_field")
+        return super().wrap(data)
+
+    @classmethod
+    def get_schema_params(cls) -> Tuple[Tuple, Dict]:
+        (schema, *other_args), kwargs = super().get_schema_params()
+        schema.pop(SchemaOptional("doc_type"))
+        old_style = schema.copy()
+        old_style.update({
+            "doc_type": "FormUserAncestorLocationField",
+            "location_field": str,
+        })
+        new_style = schema.copy()
+        new_style.update({
+            SchemaOptional("doc_type"): "FormUserAncestorLocationField",
+            "form_user_ancestor_location_field": str,
+        })
+        schema = Or(old_style, new_style)
+        return (schema, *other_args), kwargs
+
+    def get_commcare_value(self, case_trigger_info: CaseTriggerInfo) -> Any:
         user_id = case_trigger_info.form_question_values.get('/metadata/userID')
         location = get_owner_location(case_trigger_info.domain, user_id)
         if location:
-            return get_ancestor_location_metadata_value(location, self.location_field)
-
-
-class JsonPathMixin(DocumentSchema):
-    """
-    Used for importing a value from a JSON document.
-    """
-    jsonpath = StringProperty(required=True, validators=not_blank)
-
-    def _get_external_value(self, external_data):
-        jsonpath = parse_jsonpath(self.jsonpath)
-        matches = jsonpath.find(external_data)
-        values = [m.value for m in matches]
-        if not values:
-            return None
-        elif len(values) == 1:
-            return values[0]
-        else:
-            return values
-
-    def get_import_value(self, external_data):
-        external_value = self._get_external_value(external_data)
-        return self.deserialize(external_value)
+            return get_ancestor_location_metadata_value(
+                location, self.form_user_ancestor_location_field
+            )
 
 
-class JsonPathCaseProperty(CaseProperty, JsonPathMixin):
+@attr.s
+class CasePropertyConstantValue(ConstantValue, CaseProperty):
     pass
 
 
-class JsonPathCasePropertyMap(CasePropertyMap, JsonPathMixin):
-    pass
+def as_value_source(data: dict) -> ValueSource:
+    for subclass in recurse_subclasses(ValueSource):
+        try:
+            args, kwargs = subclass.get_schema_params()
+            # .. WARNING::
+            #    ``validate()`` can modify ``data``. This can break
+            #    validation for other subclasses! If it is ever
+            #    necessary to modify ``data``, it should be deep-copied
```

Comment:
I'm not sure how to reconcile "_``validate()`` can modify ``data``_" with "_If it is ever necessary to modify ``data``, ..._".

Is the phrase "_``validate()`` can modify ``data``_" stating that it is a possibility given some theoretical implementation of `validate()` that does not currently exist at the time of writing this comment? Or does it mean there are known implementations that do modify `data`? If the latter then this should be doing a deep copy now, right?

Reply:
> does it mean there are known implementations that do modify data?

That is what I thought. But I thought wrong! I reread the code, and `validate()` returns a new dictionary.

> If the latter then this should be doing a deep copy now, right?

I don't love deep copy. My reason for that (incorrect) comment was to tell a future developer to add `deepcopy()` only when they wrote a schema that modified values. I agree that adding it pre-emptively would future-proof the code. And considering all ValueSource classes are shallow and small, `deepcopy()` would be fast, so speed and resources wouldn't be an issue.

Luckily `validate()` is sensible, and doesn't modify its parameter by reference. So I've added a test to verify that, and I'm dropping the comment.

Reply:
> I don't love deep copy.

I agree. It should be avoided. I'm glad you were able to invalidate and remove that comment.